### PR TITLE
Feature/explore courses

### PR DESCRIPTION
**Pull request by KirillMakarov**
**YouTrack task**: [#APPS-1607](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1607), [#APPS-1608](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1608), [#APPS-1609](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1609), [Apps-1637](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1637)

**Description List**:
* New Catalog screen with Course Collections.
* Filter courses by Language
* Default filter is `Ru` for Russian language devices and `En` for others
* Save scrolling position on rotation in carousels for home screen
* open cached results in View all for my courses and popular
* simplify backgrounds of Buttons in `new_course_item` on Pre-Lollipop
* rewrite loading placeholders for new widgets
* handle offline mode at Catalog Screen
* fix crash on main screen [this](https://github.com/StepicOrg/stepik-android/pull/165/commits/b0567f48247be6b46e759f0fdbf4c687a8ac5fa4#diff-6d2ec6ed9fcc221abbcc95e6cc526bc4R328) for [Apps-1580](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1580)
* fetch popular courses by language

### CODE REVIEW

File: app/src/main/java/org/stepic/droid/ui/custom/LanguageTextView.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import android.support.annotation.ColorRes
+import android.support.annotation.DrawableRes
+import android.support.v7.widget.AppCompatTextView
+import android.util.AttributeSet
+import android.view.View
+import android.widget.Checkable
+import org.stepic.droid.R
+import org.stepic.droid.util.ColorUtil
+
+class LanguageTextView
```
Comment (eadm) [inline app/src/main/java/org/stepic/droid/ui/custom/LanguageTextView.kt]: Can the same behavior be achieved with styled radio buttons?

File: app/src/main/java/org/stepic/droid/ui/util/SearchHelper.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import android.support.v7.widget.SearchView
+import android.view.Menu
+import android.view.MenuInflater
+import android.view.MenuItem
+import android.widget.ImageView
+import org.stepic.droid.R
+import org.stepic.droid.ui.util.CloseIconHolder.getCloseIconDrawableRes
+
+object SearchHelper {
+    fun createSearch(menu: Menu, inflater: MenuInflater, activity: Activity): MenuItem {
```
Comment (eadm) [inline app/src/main/java/org/stepic/droid/ui/util/SearchHelper.kt]: This object is not used and doesn't use advantages of custom `SearchView`

File: app/src/main/java/org/stepic/droid/core/DefaultFilterImpl.kt [hunk 3]
```diff
@@ -41,12 +25,6 @@ class DefaultFilterImpl
...
         }
 
-        if (!needResolveLanguage()) {
-            val isNeedRussianLocal = true
-            isNeedRussian = isNeedRussianLocal
-            return isNeedRussianLocal // show russian always for old users (before 1.16)
-        }
-
         val locale: Locale =
                 if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
```
[the remark this answers is not in the document]
Comment (KirillMakarov) [reply app/src/main/java/org/stepic/droid/core/DefaultFilterImpl.kt]: It is good idea, but lint will highlight as error this expression

[the remark this answers is not in the document]
Comment (KirillMakarov) [reply app/src/main/java/org/stepic/droid/core/DefaultFilterImpl.kt]: We should rework `AndroidVersion.kt` in future

File: app/src/main/java/org/stepic/droid/ui/custom/LanguageTextView.kt (hunk 1 above)
Comment (KirillMakarov) [reply app/src/main/java/org/stepic/droid/ui/custom/LanguageTextView.kt]: May be, I didn't research enough. We can try it in future releases, when number of languages will grow.

**Pull request merged**